File: packages/core/src/tree.ts
/**
 * Tree behavior machinery.
 * Contract: docs/contracts/components/tree.md (pending — see the g11.004
 * classification sweep), "Behavior Machine" section.
 *
 * Pure logic for the tree: visible-row flattening (drives keyboard order,
 * range selection, and virtualization), tri-state checkbox cascade,
 * keyboard-intent resolution, shift-range selection, sibling-reorder
 * targets, and virtual-scroll windowing. Rename drafts, drag DOM plumbing,
 * and focus calls stay adapter-side.
 */

import {
  dropCommitDestination,
  type DragDropCommitResult,
  type DropEligibility,
  type DropIntent,
} from "./drag-drop";

export interface TreeNodeLike {
  value: string;
  children?: TreeNodeLike[];
  isBranch?: boolean;
  isDisabled?: boolean;
}

export interface TreeRow<T extends TreeNodeLike = TreeNodeLike> {
  node: T;
  depth: number;
  parent: string | null;
}

export function isTreeBranch(node: TreeNodeLike): boolean {
  return Boolean(node.isBranch) || (node.children?.length ?? 0) > 0;
}

/** Depth-first flatten of the rows currently visible given the expansion set. */
export function flattenVisibleTreeRows<T extends TreeNodeLike>(
  nodes: readonly T[],
  expandedValues: readonly string[],
): TreeRow<T>[] {
  const rows: TreeRow<T>[] = [];

  const walk = (list: readonly T[], depth: number, parent: string | null): void => {
    for (const node of list) {
      rows.push({ node, depth, parent });

      if (isTreeBranch(node) && expandedValues.includes(node.value) && node.children?.length) {
        walk(node.children as T[], depth + 1, node.value);
      }
    }
  };

  walk(nodes, 0, null);

  return rows;
}

export function findTreeNode<T extends TreeNodeLike>(nodes: readonly T[], value: string): T | null {
  for (const node of nodes) {
    if (node.value === value) {
      return node;
    }

    const found = node.children ? findTreeNode(node.children as T[], value) : null;

    if (found) {
      return found;
    }
  }

  return null;
}

// ── Tri-state checkbox cascade ──

/** Checkable atoms under a node: itself when childless, else every leaf descendant. */
export function treeCheckableUnder(node: TreeNodeLike): string[] {
  if (!node.children?.length) {
    return [node.value];
  }

  return node.children.flatMap(treeCheckableUnder);
}

export type TreeCheckState = "checked" | "unchecked" | "mixed";

export function treeCheckState(node: TreeNodeLike, checkedValues: readonly string[]): TreeCheckState {
  const leaves = treeCheckableUnder(node);
  const checked = leaves.filter((value) => checkedValues.includes(value)).length;

  return checked === 0 ? "unchecked" : checked === leaves.length ? "checked" : "mixed";
}

/** Cascade toggle: all-on clears the subtree, otherwise checks it fully. */
export function treeToggleCheck(node: TreeNodeLike, checkedValues: readonly string[]): string[] {
  const leaves = treeCheckableUnder(node);
  const allOn = leaves.every((value) => checkedValues.includes(value));

  return allOn
    ? checkedValues.filter((value) => !leaves.includes(value))
    : [...new Set([...checkedValues, ...leaves])];
}

// ── Shift-range selection over the visible order ──

export function treeRangeSelection<T extends TreeNodeLike>(
  rows: readonly TreeRow<T>[],
  anchorValue: string | null,
  toValue: string,
): string[] | null {
  const order = rows.map((row) => row.node.value);
  const a = order.indexOf(anchorValue ?? toValue);
  const b = order.indexOf(toValue);

  if (a === -1 || b === -1) {
    return null;
  }

  const [lo, hi] = a <= b ? [a, b] : [b, a];

  return order
    .slice(lo, hi + 1)
    .filter((value) => !rows.find((row) => row.node.value === value)?.node.isDisabled);
}

// ── Sibling reorder target ──

export function treeSiblingReorderTarget<T extends TreeNodeLike>(
  siblings: readonly T[],
  value: string,
  direction: 1 | -1,
): { target: string; position: "before" | "after" } | null {
  const index = siblings.findIndex((node) => node.value === value);
  const nextIndex = index + direction;

  if (index < 0 || nextIndex < 0 || nextIndex >= siblings.length) {
    return null;
  }

  const target = siblings[nextIndex];

  if (target === undefined) {
    return null;
  }

  return { target: target.value, position: direction < 0 ? "before" : "after" };
}

// ── Keyboard intents ──

export type TreeKeyIntent =
  | { type: "focus"; value: string | null; extendSelection: boolean }
  | { type: "expand"; value: string }
  | { type: "collapse"; value: string }
  | { type: "focusParent"; parent: string | null }
  | { type: "moveSibling"; direction: 1 | -1 }
  | { type: "activate" }
  | { type: "toggleSelection" }
  | { type: "startRename" };

/**
 * Resolve a treeitem keydown to an intent the adapter executes. Mirrors the
 * pre-machine Svelte behavior exactly, including Alt+Arrow reorder and
 * Shift+Arrow range extension over enabled rows.
 */
export function treeKeydownIntent<T extends TreeNodeLike>(
  rows: readonly TreeRow<T>[],
  currentValue: string,
  key: string,
  modifiers: { altKey: boolean; shiftKey: boolean },
  options: { reorderable: boolean; expandedValues: readonly string[] },
): TreeKeyIntent | null {
  const index = rows.findIndex((row) => row.node.value === currentValue);
  const row = rows[index];

  if (!row) {
    return null;
  }

  const node = row.node;

  switch (key) {
    case "ArrowDown": {
      if (modifiers.altKey && options.reorderable) {
        return { type: "moveSibling", direction: 1 };
      }

      const next = rows[index + 1];

      return {
        type: "focus",
        value: next?.node.value ?? null,
        extendSelection: modifiers.shiftKey && next !== undefined && !next.node.isDisabled,
      };
    }
    case "ArrowUp": {
      if (modifiers.altKey && options.reorderable) {
        return { type: "moveSibling", direction: -1 };
      }

      const prev = rows[index - 1];

      return {
        type: "focus",
        value: prev?.node.value ?? null,
        extendSelection: modifiers.shiftKey && prev !== undefined && !prev.node.isDisabled,
      };
    }
    case "ArrowRight": {
      if (!isTreeBranch(node)) {
        return null;
      }

      if (!options.expandedValues.includes(node.value)) {
        return { type: "expand", value: node.value };
      }

      return { type: "focus", value: rows[index + 1]?.node.value ?? null, extendSelection: false };
    }
    case "ArrowLeft": {
      if (isTreeBranch(node) && options.expandedValues.includes(node.value)) {
        return { type: "collapse", value: node.value };
      }

      return { type: "focusParent", parent: row.parent };
    }
    case "Home":
      return { type: "focus", value: rows[0]?.node.value ?? null, extendSelection: false };
    case "End":
      return { type: "focus", value: rows[rows.length - 1]?.node.value ?? null, extendSelection: false };
    case "Enter":
      return node.isDisabled ? null : { type: "activate" };
    case " ":
      return node.isDisabled ? null : { type: "toggleSelection" };
    case "F2":
      return { type: "startRename" };
    default:
      return null;
  }
}

// ── Virtual-scroll windowing ──

export interface TreeVirtualWindow {
  startIndex: number;
  endIndex: number;
  offsetY: number;
  totalHeight: number;
}

export function treeVirtualWindow(
  rowCount: number,
  rowHeightPx: number,
  scrollTop: number,
  viewportHeightPx: number,
  overscan = 6,
  pinIndex: number | null = null,
): TreeVirtualWindow {
  let startIndex = Math.max(0, Math.floor(scrollTop / rowHeightPx) - overscan);
  let endIndex = Math.min(rowCount, Math.ceil((scrollTop + viewportHeightPx) / rowHeightPx) + overscan);

  if (pinIndex !== null && pinIndex >= 0 && pinIndex < rowCount) {
    if (pinIndex < startIndex) startIndex = pinIndex;
    if (pinIndex >= endIndex) endIndex = pinIndex + 1;
  }

  return {
    startIndex,
    endIndex,
    offsetY: startIndex * rowHeightPx,
    totalHeight: rowCount * rowHeightPx,
  };
}

/** True when `descendant` is in the subtree of `ancestor` (not itself). */
export function treeSubtreeContains<T extends TreeNodeLike>(
  nodes: readonly T[],
  ancestor: string,
  descendant: string,
): boolean {
  const node = findTreeNode(nodes, ancestor);
  if (!node || ancestor === descendant) return false;
  return findTreeNode(node.children ?? [], descendant) !== null;
}

/**
 * Whether a tree drop is eligible: not self, not missing, not into the source
 * subtree. Hosts still apply `reorder_nodes`; this is hover/drop-time rejection.
 */
export function treeCanAcceptDrop<T extends TreeNodeLike>(
  nodes: readonly T[],
  from: string,
  to: string,
): boolean {
  if (from === to) return false;
  if (!findTreeNode(nodes, from) || !findTreeNode(nodes, to)) return false;
  return !treeSubtreeContains(nodes, from, to);
}

/** Hover and commit eligibility against the authoritative destination. */
export function treeDropEligibility<T extends TreeNodeLike>(
  nodes: readonly T[],
  from: string,
  intent: DropIntent,
): DropEligibility {
  const dest = dropCommitDestination(intent);
  const destNode = findTreeNode(nodes, dest.targetId);
  if (!destNode) {
    return { accepted: false, reason: "missing" };
  }
  if (destNode.isDisabled) {
    return { accepted: false, reason: "disabled" };
  }
  if (!treeCanAcceptDrop(nodes, from, dest.targetId)) {
    return { accepted: false, reason: dest.targetId === from ? "self" : "subtree" };
  }
  return { accepted: true, intent };
}

/** Where `value` sits in the tree: its parent, sibling list, and index. */
export interface TreeLocation<T extends TreeNodeLike = TreeNodeLike> {
  parent: string | null;
  siblings: readonly T[];
  index: number;
}

export function treeLocate<T extends TreeNodeLike>(
  nodes: readonly T[],
  value: string,
): TreeLocation<T> | null {
  const search = (siblings: readonly T[], parent: string | null): TreeLocation<T> | null => {
    const index = siblings.findIndex((node) => node.value === value);
    if (index >= 0) return { parent, siblings, index };
    for (const node of siblings) {
      if (!node.children?.length) continue;
      const found = search(node.children as T[], node.value);
      if (found) return found;
    }
    return null;
  };
  return search(nodes, null);
}

export type TreeDropPosition = "before" | "after" | "inside";

export interface TreeReorderSubject {
  readonly sourceValue: string;
  readonly movingValues: readonly string[];
}

export interface TreeReorderCandidate {
  readonly subject: TreeReorderSubject;
  readonly intent: DropIntent;
}

export interface TreeReorderAuthority {
  projectMovingValues(
    sourceValue: string,
    selectedValues: readonly string[],
  ): readonly string[];
  canDrop(candidate: TreeReorderCandidate): DropEligibility;
  onDrop(
    candidate: TreeReorderCandidate,
  ): DragDropCommitResult | Promise<DragDropCommitResult>;
}

export type TreeReorderProps =
  | {
      reorderAuthority?: null;
      onReorder?: (from: string, to: string, position: TreeDropPosition) => void;
    }
  | { reorderAuthority: TreeReorderAuthority; onReorder?: never };

export function isTreeDropPosition(position: DropIntent["position"]): position is TreeDropPosition {
  return position === "before" || position === "after" || position === "inside";
}

/**
 * True when a projected moving set is non-empty, unique, contains the source,
 * and every value exists in the current tree. Invalid sets are refused, never
 * reduced to `[sourceValue]`.
 */
export function treeMovingValuesAreValid<T extends TreeNodeLike>(
  nodes: readonly T[],
  sourceValue: string,
  movingValues: readonly string[],
): boolean {
  if (movingValues.length === 0) return false;
  const seen = new Set<string>();
  for (const value of movingValues) {
    if (seen.has(value)) return false;
    seen.add(value);
    if (findTreeNode(nodes, value) === null) return false;
  }
  return seen.has(sourceValue);
}

/** Latch a valid projected subject, or `null` when the projection is refused. */
export function treeLatchReorderSubject<T extends TreeNodeLike>(
  nodes: readonly T[],
  sourceValue: string,
  movingValues: readonly string[],
): TreeReorderSubject | null {
  if (!treeMovingValuesAreValid(nodes, sourceValue, movingValues)) return null;
  return {
    sourceValue,
    movingValues: Object.freeze([...movingValues]),
  };
}

/**
 * Generic Tree safety, then live host policy, then structural validation of
 * any rewritten destination. Hover fields stay on the accepted intent; only
 * `destination` may change.
 */
export function treeAuthorityDropEligibility<T extends TreeNodeLike>(
  nodes: readonly T[],
  subject: TreeReorderSubject | null,
  authority: TreeReorderAuthority | null | undefined,
  intent: DropIntent,
): DropEligibility {
  if (!subject || !authority) {
    return { accepted: false, reason: "unavailable" };
  }
  for (const moving of subject.movingValues) {
    const safety = treeDropEligibility(nodes, moving, intent);
    if (!safety.accepted) return safety;
  }
  const policy = authority.canDrop({ subject, intent });
  if (!policy.accepted) return policy;
  const dest = dropCommitDestination(policy.intent);
  if (!isTreeDropPosition(dest.position) || !isTreeDropPosition(intent.position)) {
    return { accepted: false, reason: "unavailable" };
  }
  const acceptedIntent: DropIntent = {
    targetId: intent.targetId,
    position: intent.position,
    operation: intent.operation,
    destination: dest,
  };
  for (const moving of subject.movingValues) {
    const safety = treeDropEligibility(nodes, moving, acceptedIntent);
    if (!safety.accepted) return safety;
  }
  return { accepted: true, intent: acceptedIntent };
}

File: packages/contracts/headless/src/tree.rs
//! Tree behavior machinery. Mirror of core `tree.ts`: visible-row
//! flattening, tri-state checkbox cascade, keyboard intents, shift-range
//! selection, sibling-reorder targets, and virtual-scroll windowing.
//! Generic over the host's node type via [`TreeNodeLike`] so `poodle-specs`
//! and runtime hosts can delegate without converting their trees.

pub trait TreeNodeLike: Sized {
    fn value(&self) -> &str;
    fn children(&self) -> &[Self];
    fn is_branch_flag(&self) -> bool;
    fn is_disabled(&self) -> bool;
}

pub fn is_tree_branch<N: TreeNodeLike>(node: &N) -> bool {
    node.is_branch_flag() || !node.children().is_empty()
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TreeRow {
    pub value: String,
    pub depth: usize,
    pub parent: Option<String>,
    pub disabled: bool,
    pub branch: bool,
    pub expanded: bool,
}

/// Depth-first flatten of the rows currently visible given the expansion set.
pub fn flatten_visible_tree_rows<N: TreeNodeLike>(nodes: &[N], expanded: &[String]) -> Vec<TreeRow> {
    fn walk<N: TreeNodeLike>(
        nodes: &[N],
        expanded: &[String],
        depth: usize,
        parent: Option<&str>,
        out: &mut Vec<TreeRow>,
    ) {
        for node in nodes {
            let branch = is_tree_branch(node);
            let is_expanded = expanded.iter().any(|value| value == node.value());

            out.push(TreeRow {
                value: node.value().to_string(),
                depth,
                parent: parent.map(str::to_string),
                disabled: node.is_disabled(),
                branch,
                expanded: branch && is_expanded,
            });

            if branch && is_expanded && !node.children().is_empty() {
                walk(node.children(), expanded, depth + 1, Some(node.value()), out);
            }
        }
    }

    let mut out = Vec::new();
    walk(nodes, expanded, 0, None, &mut out);
    out
}

pub fn find_tree_node<'a, N: TreeNodeLike>(nodes: &'a [N], value: &str) -> Option<&'a N> {
    for node in nodes {
        if node.value() == value {
            return Some(node);
        }

        if let Some(found) = find_tree_node(node.children(), value) {
            return Some(found);
        }
    }

    None
}

// ── Tri-state checkbox cascade ──

/// Checkable atoms under a node: itself when childless, else every leaf descendant.
pub fn tree_checkable_under<N: TreeNodeLike>(node: &N) -> Vec<String> {
    if node.children().is_empty() {
        return vec![node.value().to_string()];
    }

    node.children().iter().flat_map(tree_checkable_under).collect()
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum TreeCheckState {
    Checked,
    Unchecked,
    Mixed,
}

pub fn tree_check_state<N: TreeNodeLike>(node: &N, checked_values: &[String]) -> TreeCheckState {
    let leaves = tree_checkable_under(node);
    let checked = leaves.iter().filter(|value| checked_values.contains(value)).count();

    if checked == 0 {
        TreeCheckState::Unchecked
    } else if checked == leaves.len() {
        TreeCheckState::Checked
    } else {
        TreeCheckState::Mixed
    }
}

/// Cascade toggle: all-on clears the subtree, otherwise checks it fully.
pub fn tree_toggle_check<N: TreeNodeLike>(node: &N, checked_values: &[String]) -> Vec<String> {
    let leaves = tree_checkable_under(node);
    let all_on = leaves.iter().all(|value| checked_values.contains(value));

    if all_on {
        checked_values
            .iter()
            .filter(|value| !leaves.contains(value))
            .cloned()
            .collect()
    } else {
        let mut next = checked_values.to_vec();

        for leaf in leaves {
            if !next.contains(&leaf) {
                next.push(leaf);
            }
        }

        next
    }
}

// ── Shift-range selection over the visible order ──

pub fn tree_range_selection(rows: &[TreeRow], anchor: Option<&str>, to_value: &str) -> Option<Vec<String>> {
    let order: Vec<&str> = rows.iter().map(|row| row.value.as_str()).collect();
    let a = order.iter().position(|value| *value == anchor.unwrap_or(to_value))?;
    let b = order.iter().position(|value| *value == to_value)?;
    let (lo, hi) = if a <= b { (a, b) } else { (b, a) };

    Some(
        rows[lo..=hi]
            .iter()
            .filter(|row| !row.disabled)
            .map(|row| row.value.clone())
            .collect(),
    )
}

// ── Sibling reorder target ──

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TreeReorderMove {
    pub target: String,
    pub before: bool,
}

pub fn tree_sibling_reorder_target(siblings: &[String], value: &str, up: bool) -> Option<TreeReorderMove> {
    let index = siblings.iter().position(|candidate| candidate == value)?;
    let next_index = if up { index.checked_sub(1)? } else { index + 1 };
    let target = siblings.get(next_index)?;

    Some(TreeReorderMove { target: target.clone(), before: up })
}

// ── Keyboard intents ──

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum TreeKeyIntent {
    Focus { value: Option<String>, extend_selection: bool },
    Expand { value: String },
    Collapse { value: String },
    FocusParent { parent: Option<String> },
    MoveSibling { up: bool },
    Activate,
    ToggleSelection,
    StartRename,
}

pub struct TreeKeyModifiers {
    pub alt: bool,
    pub shift: bool,
}

/// Resolve a treeitem keydown to an intent the host executes. Mirrors the
/// TS core (and the pre-machine Svelte behavior) exactly.
pub fn tree_keydown_intent(
    rows: &[TreeRow],
    current_value: &str,
    key: &str,
    modifiers: TreeKeyModifiers,
    reorderable: bool,
    expanded: &[String],
) -> Option<TreeKeyIntent> {
    let index = rows.iter().position(|row| row.value == current_value)?;
    let row = &rows[index];

    match key {
        "ArrowDown" => {
            if modifiers.alt && reorderable {
                return Some(TreeKeyIntent::MoveSibling { up: false });
            }

            let next = rows.get(index + 1);

            Some(TreeKeyIntent::Focus {
                value: next.map(|r| r.value.clone()),
                extend_selection: modifiers.shift && next.map_or(false, |r| !r.disabled),
            })
        }
        "ArrowUp" => {
            if modifiers.alt && reorderable {
                return Some(TreeKeyIntent::MoveSibling { up: true });
            }

            let prev = index.checked_sub(1).and_then(|i| rows.get(i));

            Some(TreeKeyIntent::Focus {
                value: prev.map(|r| r.value.clone()),
                extend_selection: modifiers.shift && prev.map_or(false, |r| !r.disabled),
            })
        }
        "ArrowRight" => {
            if !row.branch {
                return None;
            }

            if !expanded.iter().any(|value| *value == row.value) {
                return Some(TreeKeyIntent::Expand { value: row.value.clone() });
            }

            Some(TreeKeyIntent::Focus {
                value: rows.get(index + 1).map(|r| r.value.clone()),
                extend_selection: false,
            })
        }
        "ArrowLeft" => {
            if row.branch && expanded.iter().any(|value| *value == row.value) {
                return Some(TreeKeyIntent::Collapse { value: row.value.clone() });
            }

            Some(TreeKeyIntent::FocusParent { parent: row.parent.clone() })
        }
        "Home" => Some(TreeKeyIntent::Focus {
            value: rows.first().map(|r| r.value.clone()),
            extend_selection: false,
        }),
        "End" => Some(TreeKeyIntent::Focus {
            value: rows.last().map(|r| r.value.clone()),
            extend_selection: false,
        }),
        "Enter" => {
            if row.disabled {
                None
            } else {
                Some(TreeKeyIntent::Activate)
            }
        }
        " " => {
            if row.disabled {
                None
            } else {
                Some(TreeKeyIntent::ToggleSelection)
            }
        }
        "F2" => Some(TreeKeyIntent::StartRename),
        _ => None,
    }
}

// ── Virtual-scroll windowing ──

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct TreeVirtualWindow {
    pub start_index: usize,
    pub end_index: usize,
    pub offset_y: f64,
    pub total_height: f64,
}

pub fn tree_virtual_window(
    row_count: usize,
    row_height_px: f64,
    scroll_top: f64,
    viewport_height_px: f64,
    overscan: usize,
) -> TreeVirtualWindow {
    let start_index = ((scroll_top / row_height_px).floor() as i64 - overscan as i64).max(0) as usize;
    let end_index = (((scroll_top + viewport_height_px) / row_height_px).ceil() as usize + overscan).min(row_count);

    TreeVirtualWindow {
        start_index,
        end_index,
        offset_y: start_index as f64 * row_height_px,
        total_height: row_count as f64 * row_height_px,
    }
}
